src/lib/tls/tls12/tls_handshake_state.cpp

```diff
@@ -374,13 +386,36 @@ Handshake_State::parse_sig_format(const Public_Key& key,
 
    const std::string hash_algo = hash_function_of_scheme(scheme);
 
+   // RFC 8446 4.4.3:
+   //   The SHA-1 algorithm MUST NOT be used in any signatures of
+   //   CertificateVerify messages.
+   if(scheme == Signature_Scheme::RSA_PKCS1_SHA1
+         || scheme == Signature_Scheme::ECDSA_SHA1
+         || scheme == Signature_Scheme::DSA_SHA1)
+      {
+      throw TLS_Exception(Alert::ILLEGAL_PARAMETER, "SHA-1 algorithm must not be used");
+      }
+
    if(!supported_algos_include(supported_algos, key_type, hash_algo))
       {
       throw TLS_Exception(Alert::ILLEGAL_PARAMETER,
                           "TLS signature extension did not allow for " +
                           key_type + "/" + hash_algo + " signature");
       }
 
+
+   // RFC 8446 4.4.3:
+   //   RSA signatures MUST use an RSASSA-PSS algorithm, regardless of whether
+   //   RSASSA-PKCS1-v1_5 algorithms appear in "signature_algorithms".
+   if(version() == Protocol_Version::TLS_V13 && key_type == "RSA" &&
+         (scheme == Signature_Scheme::RSA_PKCS1_SHA1
+          || scheme == Signature_Scheme::RSA_PKCS1_SHA256
+          || scheme == Signature_Scheme::RSA_PKCS1_SHA384
+          || scheme == Signature_Scheme::RSA_PKCS1_SHA512))
+      {
+      throw TLS_Exception(Alert::ILLEGAL_PARAMETER, "RSA signatures must use an RSASSA-PSS algorithm");
+      }
```

Comment:
This hunk should be removed as `version() == Protocol_Version::TLS_V13` won't ever be true. This is a left-over from when the premature TLS 1.3 code still depended on this handshake state implementation.

Reply:
This got replaced by `Signature_Scheme::is_compatible_with`.